pyqtgraph/graphicsItems/PlotDataItem.py

```diff
@@ -506,21 +512,34 @@ def setData(self, *args, **kargs):
                 #self.opts[k] = kargs[k]
             #scatterArgs[v] = self.opts[k]
 
-        if y is None or len(y) == 0: # empty data is represented as None
-            self.yData = None
-        else: # actual data is represented by ndarray
-            if not isinstance(y, np.ndarray):
-                y = np.array(y)
-            self.yData = y.view(np.ndarray)
-            if x is None:
-                x = np.arange(len(y))
-                
-        if x is None or len(x)==0: # empty data is represented as None
-            self.xData = None
-        else: # actual data is represented by ndarray
-            if not isinstance(x, np.ndarray):
-                x = np.array(x)
-            self.xData = x.view(np.ndarray)  # one last check to make sure there are no MetaArrays getting by
+        if x is not None:
+            if len(x) == 0:
+                x = None # represent empty data as None
+            elif not isinstance(x, np.ndarray):
+                x = np.array(x) # convert to ndarray
+            else:
+                x = x.view(np.ndarray) # last check to remove MetaArrays
+
+        if y is not None:
+            if len(y) == 0:
+                y = None # represent empty data as None
+            elif not isinstance(y, np.ndarray):
+                y = np.array(y) # convert to ndarray
+            else:
+                y = y.view(np.ndarray) # last check to remove MetaArrays
+
+        if x is None and y is not None: # generate x indices if no values are given
+            x = np.arange( float(len(y)) )
+            if append and self.xData is not None: # ...then continue after the latest element
+                x += self.xData[-1]+1
+
+        if not append or self.yData is None:
+            self.xData = x # set to new data
+            self.yData = y
+        elif y is not None: # append; note that x cannot be None if y is given
+            self.xData = np.append(self.xData, x)
```

Comment:
Hmm. `np.append()` is traditionally known to be very slow (due to additional copying and re-allocation of mem) so I'm wondering if you've actually noticed any improvements with this on fast updates @NilsNemitz?

Here's some resources to back that claim up:
- https://stackoverflow.com/a/25649863
- https://stackoverflow.com/a/46103554
- https://numpy.org/doc/stable/reference/generated/numpy.append.html#numpy.append

I would somewhat anticipate a pre-allocated large(r) array that can be passed in and instead the "appending" is done through creating new graphics using [`QPainterPath.addPath()`](https://doc.qt.io/qt-5/qpainterpath.html#addPath) and a "diff" with whatever the old rendered path was and the new input array. This will likely require a manual call to `pyqtgraph.functions.arrayToQpath()`  to do the updating from new data and generate the new path segments.

To me, if the *data* is being appended (or *updated*) that should be up the client code using `pyqtgraph` and and `PlotCurveItem` should be in charge of just figuring out what new graphics need to be generated and displayed.

I'll follow up with a more in depth commentary and examples hopefully by tonight.

Reply:
Hi @goodboy! Nice to see somebody digging into the code to work on increased performance.

This PR just aims to implement what was attempted in earlier PR #850 , with maybe slightly less redundant code.
The envisioned use case is an extremely simple, Labview-like code path for extremely simple projects like periodically reading a voltage from a multimeter and adding it to an existing graph. Since data can be saved from the context menu later, this seemed useful enough to throw some minimally invasive lines of code at it.

I am fully aware that the solution here is slow. I agree that adding a small subset of data can in principle be faster than replacing the data altogether, but getting that to work without breaking other functionality (which expects numpy arrays of data in xData and yData) was a larger project than I felt ready to tackle.

I will be happy to see a high performance solution to replace this. Even happier if it plays nicely with the `PlotDataItem` data mangling (e.g. PR #1140), which seems needed to get Qt to draw some of my data sets properly. :)

I consider manual messing with QPainterPath outside my area of expertise. But I am happy to hear your suggestions on this PR, or to contribute to your more in-depth effort.